File: Data Structures & Algorithms/products-of-array-discluding-self/submission-7.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        1 , 2, 4, 6
        """
        N = len(nums)
        res = [1 for i in range(N)]

        prefix = 1
        for idx in range(N):
            res[idx] = prefix
            prefix *= nums[idx]

        postfix = 1
        for idx in range(N-1 , -1, -1):
            res[idx] *= postfix
            postfix *= nums[idx]

        return res

        

File: Data Structures & Algorithms/top-k-elements-in-list/submission-7.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        List<Integer>[] bucket = new List[nums.length+1];
        HashMap<Integer, Integer> counts = new HashMap<>();

        for(int num: nums){
            counts.put(num, counts.getOrDefault(num, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry : counts.entrySet()) {
            int key = entry.getValue();
            if(bucket[key]==null) bucket[key] = new ArrayList<>();
            bucket[key].add(entry.getKey());
        }

        int res[] = new int[k];
        int idx = 0;
        for(int i = bucket.length-1; i >= 0; i--){
            if(bucket[i] != null){
                for(int num : bucket[i]){
                    res[idx++] = num;
                    if(idx==k){
                        return res;
                    }
                }
            }
        }

        return res;

    }
}
